java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -96,93 +142,211 @@ public void onTimer(
+        final Map<String, Integer> leftIdxMap = state.getLeftTimestampIdx();
+
+        for (AggregationFieldsDescriptor.AggregationFieldDescriptor descriptor :
+                aggregationFieldsDescriptor.getAggFieldDescriptors()) {
+            Object accumulator = accumulators.getField(descriptor.outFieldName);
+            if (accumulator == null) {
+                accumulator = descriptor.aggFunc.createAccumulator();
+                accumulators.setField(descriptor.outFieldName, accumulator);
+            }
+            if (rowToAdd != null) {
+                descriptor.aggFunc.add(
+                        accumulator, rowToAdd.getField(descriptor.inFieldName), timestamp);
             }
-            Row curRow = state.timestampToRow.get(rowTime);
 
-            // TODO: Optimize by supporting retract value from aggregation function.
-            for (AggregationFieldsDescriptor.AggregationFieldDescriptor descriptor :
-                    aggregationFieldsDescriptor.getAggFieldDescriptors()) {
-                if (timestamp - descriptor.windowSizeMs >= rowTime) {
-                    continue;
+            // Advance left idx and retract values
+            int leftIdx = leftIdxMap.getOrDefault(descriptor.outFieldName, 0);
+            for (; leftIdx < timestampListSize; ++leftIdx) {
+                long rowTime = timestampList.get(leftIdx);
+                if (timestamp - descriptor.windowSizeMs < rowTime) {
+                    if (rowTime <= timestamp) {
+                        hasRow = true;
+                    }
+                    break;
                 }
-                descriptor.aggFunc.aggregate(curRow.getField(descriptor.inFieldName), rowTime);
-                hasRow = true;
+                Row curRow = state.timestampToRow.get(rowTime);
+                descriptor.aggFunc.retract(accumulator, curRow.getField(descriptor.inFieldName));
             }
+            leftIdxMap.put(descriptor.outFieldName, leftIdx);
+
+            outputRow.setField(descriptor.outFieldName, descriptor.aggFunc.getResult(accumulator));
         }
+        state.updateLeftTimestampIdx(leftIdxMap);
+        state.updateAccumulators(accumulators);
+        outputRow.setField(rowTimeFieldName, Instant.ofEpochMilli(timestamp));
 
-        state.pruneRow(
-                ctx.getCurrentKey(), timestamp - aggregationFieldsDescriptor.getMaxWindowSizeMs());
+        state.pruneRow(timestamp, aggregationFieldsDescriptor);
 
         if (!hasRow) {
-            // output nothing if no row is aggregated.
+            state.maxRegisteredTimer.clear();
+            if (expiredRowHandler == null) {
+                // output nothing if no row is aggregated.
+                return;
+            }
+            final Row lastRow = state.lastOutputRow.value();
+            if (lastRow == null) {
+                return;
+            }
+            expiredRowHandler.handleExpiredRow(out, lastRow, timestamp);
```

Comment:
We are not emitting the last row here. The `lastRow` is passed to the `PostSlidingWindowExpiredRowHandler#handleExpiredRow`. The current implementation that outputs the default row derives the key from the expired row and added to the default row. 
And in the future, we will add an implementation that sends out retract message when the last row expired and it needs the expired row to send out the retract message.

Reply:
I see. The existing doc of `PostSlidingWindowDefaultRowExpiredRowHandler` says "PostSlidingWindowDefaultRowExpiredRowHandler output a row with the default values when a row is expired". Can you update it to mention that it uses the key from the expired row?

Also, instead of saying "default value", it might be more informative to mention that we use the zero-valued row etc. And maybe rename `defaultRow` to `zeroValuedRow` as appropriate in this PR.